File: IAIso-v5.0/core/iaiso-swift/Sources/IAIsoAudit/JSONLFileSink.swift
import Foundation

/// Appends events as JSONL to a file. I/O errors are silently dropped.
public final class JSONLFileSink: Sink, @unchecked Sendable {
    private let lock = NSLock()
    private let url: URL

    public init(path: String) {
        self.url = URL(fileURLWithPath: path)
    }

    public init(url: URL) {
        self.url = url
    }

    public func emit(_ event: Event) {
        let line = event.toJSON() + "\n"
        guard let bytes = line.data(using: .utf8) else { return }
        lock.lock()
        defer { lock.unlock() }
        do {
            if FileManager.default.fileExists(atPath: url.path) {
                let handle = try FileHandle(forWritingTo: url)
                defer { try? handle.close() }
                try handle.seekToEnd()
                try handle.write(contentsOf: bytes)
            } else {
                try bytes.write(to: url, options: .atomic)
            }
        } catch {
            // Silently drop — sinks must not propagate I/O errors.
        }
    }
}

File: IAIso-v5.0/core/iaiso-ruby/lib/iaiso/audit/jsonl_file_sink.rb
# frozen_string_literal: true

require "monitor"

module IAIso
  module Audit
    # Appends events as JSONL to a file.
    # I/O errors are silently dropped — sinks must not propagate them.
    class JSONLFileSink
      include MonitorMixin

      def initialize(path)
        super()
        @path = path
      end

      def emit(event)
        line = event.to_json + "\n"
        synchronize do
          File.open(@path, "a:UTF-8") { |f| f.write(line) }
        rescue StandardError
          # swallow
        end
      end
    end
  end
end
